### ISSUE TITLE
No sound effects when using D1X-Rebirth in macOS with MacPlay CD release of Descent

### ISSUE THREAD

**Issue by almeath** [dxx-rebirth/dxx-rebirth#630]
I do not know if this is a bug, or expected behavior. I cannot find reference to it anywhere. 

Hopefully the title of the issue is self-explanatory. While I can get the MacPlay CD version of Descent to work flawlessly in terms of its graphics and CD soundtrack (ripped FLAC), I get absolutely no sound effects when using the PIG file from the the CD installation. I tested two different copies of the CD release to make sure it was not a fault with the original media.

When I try substituting the Mac PIG file with the version from the PC release, it appears to work. I have tested the first level and it all seems good, but I do not know if I am likely to experience further issues with this work-around. The Mac PIG file is reported on my system as about 1MB smaller in size than the PC version.

I have built by own DXX-Rebirth app, and also tried using a pre-compiled version from Mac Source Ports. I get the same result with both, so I know I did not mess up my build in any way.

I also see the same terminal output when using either PIG file. No apparent errors. I see a message about using SDL_mixer library v1.2.12.

Is the Mac PIG file not compatible with DXX-Rebirth?

**Comment by vLKp** [comment]
> I see a message about using SDL_mixer library v1.2.12.

This is an informational message only.  All recent builds log this type of data, so that if a bug in a supporting library is suspected, the version of that library can be checked against any lists of known issues.  I am not aware of any known issues with SDL_mixer that would lead to the reported issue.

> Is the Mac PIG file not compatible with DXX-Rebirth?

As far as I know, the Mac PIG _should_ work, and there is definitely special code to handle it.  However, I do not have this CD, so I cannot check whether it does work, nor can I readily debug it if, as you report, it does not work as-is.  I expect from your post that the Mac PIG is one of the copyright-restricted files, and there are no legal free downloads of it.  Is that correct?  If that is incorrect and there is a legal downloadable version of it, please link to it so that I can retrieve a copy.  If it is not available for legal free download, you will need to debug this yourself or find someone who has this file and can debug it for you.  I can advise where to put debugging statements if you wish to debug this yourself.

**Comment by almeath** [comment]
There is no sign of the Mac version keeping its audio files anywhere else, as far as I can tell. The application/executable is 1.8 MB in size. Apart from files called "CHAOS.HOG and "CHAOS.TXT" only the HOG and PID files are installed in the data folder.

The PID file with the demo is 2.7 MB and the one with the full CD version is about 3.9 MB, whereas the PC version is about 4.9 MB. Neither of the Mac versions, demo or full, will produce any sound, whether matched up with the demo or full CD HOG file. 

Unfortunately, I do not have original MacOS classic hardware to test this on. Emulation under SheepShaver fails because as soon as any 3D element needs to be displayed (either in the intro sequence or once the first level starts, I get an entirely black screen). It seems to be a flaw in the emulation software. I am quite certain that the files or media are not damaged, as the original CDs appear flawless and the music tracks play properly.

**Comment by vLKp** [comment]
What is the size of the Mac PIG file that fails to produce sound?  I did not expect that the CDs were defective, but I suspect that there is some difference in how the Mac PIG file is constructed, as compared to the PC PIG file, that causes Rebirth to fail to use it.  If we understood how it was different, it should be possible to fix it.

**Comment by almeath** [comment]
The exact size of the PIG file that comes on the Decent CD is 3,975,533 bytes with an MD5 checksum of 80b5a00803350920eee681abe2f354a1.

For the PIG from the demo, it is a smaller size of 2,714,487 bytes with an MD5 checksum of f08c77ce55f89f3b876a25b19c1c08de.

I have tested both, to ensure that the HOG and PIG from the demo behave in the same way as the the ones from the full game. Using the PC PIG file with the demo's HOG does allow the sound effects to work.

Therefore, if we are dealing with structural differences in the contents of the PIG, resolving it for the demo should also resolve it for the full game. 

As the demo is freeware, I can provide the HOG and PIG files from that here:

[Data.zip](https://github.com/dxx-rebirth/dxx-rebirth/files/8490593/Data.zip)
.

**Comment by almeath** [comment]
This appears to have been a known issue for a long time, as I found a set of instructions from 2015 for the D2X-XL project, which mentions using the PC PIG file in place of the Mac version, to ensure that sound effects work.

**Comment by vLKp** [comment]
That is unfortunate, but it makes some sense.  I have not recently changed any code that should have broken this.  Did those instructions explain why the Mac version was inadequate?

**Comment by almeath** [comment]
I just found out something very interesting. Within an emulated MacOS System 7 environment, I used the program ResEdit to open the Descent application. I found out that the sounds are actually stored inside the Mac application in the AIF format. There are apparently 97 sound effects stored inside the application. I can even play them within ResEdit to confirm they are definitely the game's sound effects. I can use copy/paste to manually extract them and paste them into the System 7 scrapbook, but I need to find an old Mac classic app that will let me extract and save them as individual sound files.

I am not sure where to take this though. Would extracting the sounds from the application from the CD version of the game be breaching any copyright? Once I have extracted all the sounds, I do not know whether that is useful going forward, as I take is that DXX-Rebirth is setup to look inside the PIG files for sound. 

What I find really interesting is that they have included these apparently "dummy" PIG files all the Mac versions of Descent. The game is not actually looking inside those for the sound resources.

<img width="960" alt="Screen Shot 2022-04-17 at 6 40 23 am" src="https://user-images.githubusercontent.com/5726214/163693547-8bc41bae-15c0-4853-859d-a009de5ea9f9.png">

**Comment by almeath** [comment]
I have just checked the demo version, and there are more sounds in there, some with ".ds" format, which appear to be MIDI instruments. There are 72 sound effect files in the demo. So there are an extra 25 sound effects in the CD version that are not in the demo. I imagine it could take some time to figure out which ones are missing from the demo, unless the sounds are in the same sequence in both apps, which I have not had time to check yet.

**Comment by AlumiuN** [comment]
Ah, that would do it; Windows (and I think most Linux) filesystems don't have the concept of data and resource forks, so I don't think there's a particularly easy solution. The Mac PIG files aren't dummy files, though, the PIG files still contain textures and sprites, but it is a little odd that they chose not to include the sounds in it and took this path instead.

**Comment by almeath** [comment]
I think they did it that way because it was common at the time to use the QuickTime system extensions to play MIDI music and sound effects. When they released the CD version with the Redbook audio, it was not worth having the application coded any differently to the demo. Interestingly, they actually stripped out the MIDI DS files which could easily have been left in and only used a few MB if that. That meant the Mac CD version would only support the Redbook audio and nothing else for music.

**Comment by vLKp** [comment]
Files embedded in the demo Mac app are probably copyrighted, but the demo is presumably distributed under a license that permits redistribution without specific permission.  I don't know if that would include pulling the sound pieces of the demo app out for separate distribution.  It would probably be fine for users who have the Mac app, whether demo or full, to extract the files for local use - though that would still need Rebirth to read the extracted files.  GOG distributes the PC version, so new users should not have this problem.  Only people who still have their original Mac CD should be affected.

**Comment by almeath** [comment]
Yes, in my case I already owned both originals, but GOG is by far the easier solution for anyone wanting to use the Mac version with the PC sound file. It seems way too much trouble, to try to deconstruct the resource file and find a way to patch in the Mac support, given that the sound quality of these effects will be identical to the PC version. There is no qualitative edge here, unlike with some of the the Mac graphical elements.

As far as I can tell, there is no emulator that runs the Mac version properly, so DXX-Rebirth or original classic Mac hardware are the only way forward, for the most authentic experience anyway. Maybe later versions of emulators will fix this, but BIN/CUE support in SheepShaver is only available in an experimental and highly unstable patch. I tried and failed to get that to work. 

I can fully appreciate the PC version's SC-55 MIDI soundtrack (as I own an original SC-55 MKII), but for me, Descent will always mean the high(er)-res graphics and awesome re-mastered red book tracks on the MacPlay CD version of the game.
